README.md
MailChimp Subscribe
===================

A MailChimp API wrapper using Guzzle v6 which implements a simple list subscription interface using the MailChimp v3 
API.

By [Simon Hampel](http://hampelgroup.com/).

Installation
------------

The recommended way of installing MailChimp Subscribe is through [Composer](http://getcomposer.org):

    :::json
    {
        "require": {
            "hampel/mailchimp-subscribe": "~1.0"
        }
    }

Configuration
-------------

To use the MailChimp subscriber, you'll need to generate an API key using the MailChimp admin interface.

Log in to MailChimp and then go to `Account` by clicking on your username/avatar at the top and selecting Account from 
the menu.

Next, go to the `Extras` area and click on `API Keys`.

Click on the `Create a Key` button and give your new key a meaningful label so you know which one it is.

You'll also need to find the list ID of the mailing list you want to subscribe people to.

Click on `Lists` at the top and then click on your list name.

Click on the `Settings` submenu and select the `List name and campaign defaults` option.

Over on the right will be your List ID - make a note of this.

Usage
-----

### Initialisation ###

To start with, you'll need to create a MailChimp object which does the actual communication.

The quick way is to use the factory:

    :::php
    <?php
        use MailChimp\MailChimp;
        
        $apikey = 'YOUR API KEY HERE';
        
        $mc = MailChimp::make($apikey);

Alternatively, if you need more control, you can use the long-hand method:

    :::php
    <?php
        use GuzzleHttp\Client;
        use MailChimp\MailChimp;
        
        $apikey = 'YOU API KEY HERE';
        
        $config = MailChimp::getConfig($apikey);
        $guzzle = new Client($config);
        $mc = new MailChimp($guzzle, $apikey);

You'll also want to create a mailing list object to then work with the list:

    :::php
    <?php
        use MailChimp\MailingList;
        
        $listid = 'YOU LIST ID HERE';
        
        $list = new MailingList($mc, $listid);

We're now ready to work with our list.

### Get Member ###

You can retrieve the details of a member by calling the `getMember` method of the MailingList class:

    :::php
    <?php
        var_dump($list->getMember('user@example.com'));

The getMember function returns `false` if the email address was not found in the list.

### Get Member Status ###

Retrieve the current status of a list member by calling the `getMemberStatus` method:

    :::php
    <?php
        var_dump($list->getMemberStatus('user@example.com'));

Possible return values are: subscribed, unsubscribed, cleaned, pending - or false if the member was not found.  

### Check if Member Exists ###

You can check if a user is a member of a list with the `isMember` method, which returns a boolean result:

    :::php
    <?php
        if ($list->isMember('user@example.com'))
        {
        	echo "user is a member of our list";
        }
        else
        {
        	echo "user is not yet a member of our list";
        }

### Subscribe a Member ###

Subscribe a new email address to your list using the `subscribe` method:

    :::php
    <?php
        var_dump($list->subscribe('user@example.com'));

To have MailChimp send a confirmation to the user before adding them to your list (double opt-in), set the second
parameter to `true`: `$list->subscribe('user@example.com', true)`

You may also send an associative array of merge fields (name-value pairs) as a third parameter:

    :::php
    <?php
        var_dump($list->subscribe('user@example.com', false, ['FNAME' => 'Joe', 'LNAME' => 'Bloggs']));

Details about the subscribed user will be returned.

### Unsubscribe a Member ###

To remove a user from the mailing list, use the `unsubscribe` method:

    :::php
    <?php
        var_dump($list->unsubscribe('user@example.com'));

### Clean a Member ###

To clean an address from the mailing list, use the `clean` method:

    :::php
    <?php
        var_dump($list->clean('user@example.com'));

Cleaned addresses are addresses that have bounced too many times or are now considered invalid.


CHANGELOG.md
CHANGELOG
=========

1.0.0 (2015-11-19)
------------------

* rewrote package to work with Guzzle v6

0.2.1 (2015-06-17)
------------------

* bug fix - missing use clause for exception

0.2.0 (2015-06-17)
------------------

* refactored to split Guzzle calls into separate class so we can provide a simpler mailing list interface

0.1.0 (2015-06-17)
------------------

* initial release


src/MailChimp.php
<?php namespace MailChimp;

use Hampel\Json\Json;
use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Psr7\Response;
use GuzzleHttp\ClientInterface;
use GuzzleHttp\Exception\ParseException;
use GuzzleHttp\Exception\RequestException;
use Hampel\Json\JsonException;
use MailChimp\Exception\MailChimpParseException;
use MailChimp\Exception\MailChimpRequestException;

/**
 * The main service interface using Guzzle
 */
class MailChimp
{
	/** @var string base url for API calls */
	protected static $base_uri = 'https://<dc>.api.mailchimp.com/3.0/';

	/** @var string api key for auth */
	protected $apikey;

	/** @var Client our Guzzle HTTP Client object */
	protected $client;

	/** @var Request Psr7 Request object representing the last request made */
	protected $last_request;

	/** @var Response Guzzle Response object representing the last response from Guzzle call to MailChimp API */
	protected $last_response;

	/** @var  string a string description of the last action taken */
	protected $last_action;

	/**
	 * Constructor
	 *
	 * @param Client $client	Guzzle HTTP client
	 */
	public function __construct(ClientInterface $client, $apikey)
	{
		$this->client = $client;
		$this->apikey = $apikey;
	}

	/**
	 * Make - construct a service object
	 *
	 * @param string $api_key API Key
	 *
	 * @return MailChimp a fully hydrated MailChimp Service, ready to run
	 */
	public static function make($apikey)
	{
		return new self(new Client(self::getConfig($apikey)), $apikey);
	}

	public static function getConfig($apikey)
	{
		$dc = self::extractDc($apikey);

		return ['base_uri' => self::buildUri($dc)];
	}

	public static function buildUri($dc)
	{
		return str_replace('<dc>', $dc, self::$base_uri);
	}

	public static function extractDc($apikey)
	{
		if (empty($apikey)) return null;

		$parts = explode('-', $apikey);
		return array_pop($parts); // last part of the key is the datacenter
	}

	public function getClient()
	{
		return $this->client;
	}

	public function get($action)
	{
		$this->last_action = "GET {$action}";

		return $this->send($action, 'GET');
	}

	public function post($action, array $data = [])
	{
		$this->last_action = "POST {$action}";

		return $this->send($action, 'POST', $data);
	}

	public function patch($action, array $data = [])
	{
		$this->last_action = "PATCH {$action}";

		return $this->send($action, 'PATCH', $data);
	}

	public function send($action, $method = 'GET', $data = [], array $options = [])
	{
		if (empty($this->apikey))
		{
			throw new RuntimeException("API key not yet set");
		}

		if (!array_key_exists('auth', $options))
		{
			$options['auth'] = ['mailchimp-subscribe', $this->apikey];
		}

		$json = null;
		if (!empty($data))
		{
			try
			{
				$json = Json::encode($data);
			}
			catch (JsonException $e)
			{
				throw new RuntimeException("Could not encode JSON payload: " . $e->getMessage(), $e->getCode(), $e);
			}
		}

		$headers = [];

		$request = new Request($method, $action, $headers, $json);

		$this->last_request = $request;

		try
		{
			$response = $this->client->send($request, $options);
		}
		catch (RequestException $e)
		{
			throw new MailChimpRequestException($e->getMessage(), $e->getCode(), $e);
		}

		$this->last_response = $response;

		$body = $response->getBody();

		if ($body->getSize() > 0)
		{
			try
			{
				return Json::decode($body->getContents(), Json::DECODE_ASSOC);
			}
			catch (ParseException $e)
			{
				throw new MailChimpParseException(
					"MailChimp " . $e->getMessage() . " - last command [{$this->last_action}]",
					$e->getCode(),
					$e
				);
			}
		}
	}

	/**
	 * Return the request object from the last API call made
	 *
	 * @return Request Psr7 Request object
	 */
	public function getLastRequest()
	{
		return $this->last_request;
	}

	/**
	 * Return the response object from the last API call made
	 *
	 * @return Response Guzzle Reponse object
	 */
	public function getLastResponse()
	{
		return $this->last_response;
	}

	/**
	 * Return the status code from the last API call made
	 *
	 * @return number status code
	 */
	public function getLastStatusCode()
	{
		$last_response = $this->getLastResponse();
		if (! is_null($last_response))
		{
			return $last_response->getStatusCode();
		}
	}

	public function getLastQuery()
	{
		$last_request = $this->getLastRequest();
		if (!is_null($last_request))
		{
//			return strval(\GuzzleHttp\Psr7\Uri::resolve(\GuzzleHttp\Psr7\uri_for($this->client->getConfig('base_uri')), $last_request->getUri()));
			return strval($last_request->getUri());
		}
	}

	public function getLastAction()
	{
		return $this->last_action;
	}
}


src/MailingList.php
<?php  namespace MailChimp; 

use MailChimp\Exception\MailChimpRequestException;

class MailingList
{
	/** @var  MailChimp our MailChimp interface */
	protected $mailchimp;

	/** @var string our mailing list ID  */
	protected $list_id;

	function __construct(MailChimp $mailchimp, $list_id)
	{
		$this->mailchimp = $mailchimp;
		$this->list_id = $list_id;
	}

	public static function all(MailChimp $mailchimp)
	{
		return $mailchimp->get("lists");
	}

	public function getMember($email)
	{
		$action = "lists/{$this->list_id}/members/" . md5($email);

		try
		{
			return $this->mailchimp->get($action);
		}
		catch (MailChimpRequestException $e)
		{
			if ($e->getResponse()->getStatusCode() == '404')
			{
				return false;
			}
			else throw $e;
		}
	}

	public function getMemberStatus($email)
	{
		$member = $this->getMember($email);
		if ($member) return $member['status'];
		else return false;
	}

	public function isMember($email)
	{
		return $this->getMemberStatus($email) !== false;
	}

	public function subscribe($email, $confirm = false, $merge_fields = [])
	{
		$data = [
			'email_address' => $email,
			'status' => ($confirm ? 'pending' : 'subscribed'),
		];
		if (!empty($merge_fields)) $data['merge_fields'] = $merge_fields;

		$action = "lists/{$this->list_id}/members/";

		return $this->mailchimp->post($action, $data);
	}

	public function unsubscribe($email)
	{
		return $this->update($email, ['status' => 'unsubscribed']);
	}

	public function clean($email)
	{
		return $this->update($email, ['status' => 'cleaned']);
	}

	public function update($email, $data)
	{
		$action = "lists/{$this->list_id}/members/" . md5($email);

		return $this->mailchimp->patch($action, $data);
	}

	/**
	 * @return MailChimp
	 */
	public function getMailchimp()
	{
		return $this->mailchimp;
	}

	/**
	 * @return string
	 */
	public function getlist_id()
	{
		return $this->list_id;
	}
}


src/Subscriber.php
<?php namespace MailChimp;

use GuzzleHttp\Client;
use GuzzleHttp\Message\Request;
use GuzzleHttp\Exception\ParseException;
use GuzzleHttp\Exception\RequestException;
use MailChimp\Exception\MailChimpParseException;
use MailChimp\Exception\MailChimpRequestException;

/**
 * The main service interface using Guzzle
 */
class Subscriber
{
	/** @var string base url for API calls */
	protected static $base_uri = 'https://<dc>.api.mailchimp.com/3.0/';

	/** @var Client our Guzzle HTTP Client object */
	protected $client;

	/** @var Response Guzzle Response object representing the last response from Guzzle call to MailChimp API */
	protected $last_response;

	/** @var  string a string description of the last action taken */
	protected $last_action;

	/**
	 * Constructor
	 *
	 * @param Client $client	Guzzle HTTP client
	 */
	public function __construct(Client $client)
	{
		$this->client = $client;
	}

	/**
	 * Make - construct a service object
	 *
	 * @param string $api_key API Key
	 *
	 * @return Subscriber a fully hydrated MailChimp Service, ready to run
	 */
	public static function make($api_key)
	{
		return new self(new Client(self::getConfig($api_key)));
	}

	public static function extractDc($api_key)
	{
		if (empty($api_key)) return null;

		$parts = explode('-', $api_key);
		return array_pop($parts); // last part of the key is the datacenter
	}

	public static function buildUri($dc)
	{
		return str_replace('<dc>', $dc, self::$base_uri);
	}

	public static function getConfig($api_key)
	{
		$dc = self::extractDc($api_key);

		return [
			'base_url' => self::buildUri($dc),
			'defaults' => [
				'auth' => ['mailchimp-subscribe', $api_key]
			],
		];
	}

	public function getClient()
	{
		return $this->client;
	}

	public function getMember($listid, $email)
	{
		$action = "lists/{$listid}/members/" . md5($email);

		try
		{
			return $this->get($action);
		}
		catch (MailChimpRequestException $e)
		{
			if ($e->getResponse()->getStatusCode() == '404')
			{
				return false;
			}
			else throw $e;
		}
	}

	public function getMemberStatus($listid, $email)
	{
		$member = $this->getMember($listid, $email);
		if ($member) return $member['status'];
		else return false;
	}

	public function isMember($listid, $email)
	{
		return $this->getMemberStatus($listid, $email) !== false;
	}

	public function subscribe($listid, $email, $confirm = false, $merge_fields = array())
	{
		$data = [
			'email_address' => $email,
			'status' => ($confirm ? 'pending' : 'subscribed'),
		];
		if (!empty($merge_fields)) $data['merge_fields'] = $merge_fields;

		$action = "lists/{$listid}/members/";

		return $this->post($action, ['json' => $data]);
	}

	public function unsubscribe($listid, $email)
	{
		return $this->update($listid, $email, ['status' => 'unsubscribed']);
	}

	public function clean($listid, $email)
	{
		return $this->update($listid, $email, ['status' => 'cleaned']);
	}

	public function update($listid, $email, $data)
	{
		$action = "lists/{$listid}/members/" . md5($email);

		return $this->patch($action, ['json' => $data]);
	}

	public function get($action)
	{
		$this->last_action = "GET {$action}";

		$request = $this->client->createRequest('GET', $action);

		return $this->send($request);
	}

	public function post($action, array $data = [])
	{
		$this->last_action = "POST {$action}";

		$request = $this->client->createRequest('POST', $action, $data);

		return $this->send($request);
	}

	public function patch($action, array $data = [])
	{
		$this->last_action = "PATCH {$action}";

		$request = $this->client->createRequest('PATCH', $action, $data);

		return $this->send($request);
	}

	public function send(Request $request, array $options = [])
	{
		try
		{
			$response = $this->client->send($request, $options);
		}
		catch (RequestException $e)
		{
			throw new MailChimpRequestException($e->getMessage(), $e->getCode(), $e);
		}

		$this->last_response = $response;

		try
		{
			return $response->json();
		}
		catch (ParseException $e)
		{
			throw new MailChimpParseException("MailChimp " . $e->getMessage() . " - last command [{$this->last_action}]", $e->getCode(), $e);
		}
	}

	/**
	 * Return the response object from the last API call made
	 *
	 * @return Response Guzzle Reponse object
	 */
	public function getLastResponse()
	{
		return $this->last_response;
	}

	/**
	 * Return the status code from the last API call made
	 *
	 * @return number status code
	 */
	public function getLastStatusCode()
	{
		$last_response = $this->getLastResponse();
		if (! is_null($last_response))
		{
			return $last_response->getStatusCode();
		}
	}

	public function getLastQuery()
	{
		$last_response = $this->getLastResponse();
		if (! is_null($last_response))
		{
			return $last_response->getEffectiveUrl();
		}
	}

	public function getLastAction()
	{
		return $this->last_action;
	}
}
